pkg/backup/backup.go

```diff
@@ -501,8 +548,151 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
+				log.Error(errors.WithStack(err))
+				continue
+			}
+
+			client, err := itemBlock.itemBackupper.dynamicFactory.ClientForGroupVersionResource(gvr.GroupVersion(), resource, relatedItem.Namespace)
+			if err != nil {
+				log.Error(errors.WithStack(err))
+				continue
+			}
+
+			item, err := client.Get(relatedItem.Name, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				log.WithFields(logrus.Fields{
+					"groupResource": relatedItem.GroupResource,
+					"namespace":     relatedItem.Namespace,
+					"name":          relatedItem.Name,
+				}).Warnf("Related item was not found in Kubernetes API, can't add to item block")
+				continue
+			}
+			if err != nil {
+				log.Error(errors.WithStack(err))
+				continue
+			}
+			itemsMap[relatedItem] = append(itemsMap[relatedItem], &kubernetesResource{
+				groupResource: relatedItem.GroupResource,
+				preferredGVR:  gvr,
+				namespace:     relatedItem.Namespace,
+				name:          relatedItem.Name,
+				inItemBlock:   true,
+			})
+			log.Infof("adding %s %s/%s to ItemBlock", relatedItem.GroupResource, relatedItem.Namespace, relatedItem.Name)
+			itemBlock.AddUnstructured(relatedItem.GroupResource, item, gvr)
+			kb.executeItemBlockActions(log, item, relatedItem.GroupResource, relatedItem.Name, relatedItem.Namespace, itemsMap, itemBlock)
+		}
+	}
+}
+
+func (kb *kubernetesBackupper) backupItemBlock(itemBlock BackupItemBlock) []schema.GroupResource {
+	// find pods in ItemBlock
+	// filter pods based on whether they still need to be backed up
+	// this list will be used to run pre/post hooks
+	var postHookPods []itemblock.ItemBlockItem
+	var items []itemblock.ItemBlockItem
+	itemBlock.Log.Debug("Executing pre hooks")
+	for _, item := range itemBlock.Items {
+		if item.Gr == kuberesource.Pods {
+			skipHooks := false
+			metadata, err := meta.Accessor(item.Item)
+			if err != nil {
+				itemBlock.Log.WithError(errors.WithStack(err)).Error("Error accessing pod metadata")
+				skipHooks = true
+			}
+			// Don't run hooks if pod is excluded
+			if !skipHooks && !itemBlock.itemBackupper.itemInclusionChecks(itemBlock.Log, false, metadata, item.Item, item.Gr) {
+				skipHooks = true
+			}
+			if !skipHooks {
+				key := itemKey{
+					resource:  resourceKey(item.Item),
+					namespace: metadata.GetNamespace(),
+					name:      metadata.GetName(),
+				}
+				// Don't run hooks if pod has already been backed up
+				if _, exists := itemBlock.itemBackupper.backupRequest.BackedUpItems[key]; !exists {
+					if err := itemBlock.itemBackupper.itemHookHandler.HandleHooks(itemBlock.Log, item.Gr, item.Item, itemBlock.itemBackupper.backupRequest.ResourceHooks, hook.PhasePre, itemBlock.itemBackupper.hookTracker); err != nil {
+						itemBlock.Log.WithError(err).WithField("name", item.Item.GetName()).Error("Error running hooks for pod")
+						// if pre hook fails, flag pod as backed-up and move on
+						itemBlock.itemBackupper.backupRequest.BackedUpItems[key] = struct{}{}
+						continue
```

Comment:
The pod hook API's "fail on error" only calls for the pod itself to fail. Current Velero logic won't skip PVC, secrets, etc. either.

Reply:
Understood. We do see some problems here though they also happen with the existing code:

- The hook may be used to quiescence the volume file system or prepare the backup data to the volume
- When the hook fails and the pod is skipped, but the PVC may be still backed up
- The consequence is that users either get a corrupted backup data or empty backup data

Do you think we can enhance this in future based on the new itemBlock implementation?

Reply:
@Lyndon-Li I think we should defer this to the future, since we'll need to think carefully about the design. Do we *always* want to not restore items in the block that were returned by any IBA plugins for the pod if the pre-hook fails, or should this be part of the IBA plugin return value? If it's plugin-decided rather than optional, then we'll need to modify the IBA plugin return (with a v2 plugin). If we decide that we always want to skip related items for a pod if a hook fails, then we won't need that. But also, the model here doesn't track *which* item's IBA plugin (if any) was responsible for pulling the item into the block, so we'd need to add that too.

Also, this could introduce as many problems as it solves. If there are 2 pods in the block, either because of a shared RWX volume, or (in the future) due to the pods having volumes in the same VolumeGroup, then should one failed pod pre-hook cause all of the pods to be skipped in the backup? I think this should be a separate design discussion post-1.15.

Reply:
It may be sufficient to just fail the pod -- the user will see the PartiallyFailed backup, note that the error was on a stateful pod and therefore assume the backup is bad anyway.

Reply:
I would agree that if we decide to make a fix here in the future, it will be easier to do with the ItemBlock implementation than without.